### PR TITLE
[ENH] improvements to conditional_join for equality matches

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- minor improvement to conditional_join for equality matches
- the improvement relies on a search within the outcome of the initial binary search to avoid a full cartesian blow up - this is relevant for many to many joins, where a non-equi join (usually a range join) is expected to trim the size
- for one to one or one to many, pd.merge is used under the hood, as that is faster than binary search and the number of rows will not be greater than the larger of the two dataframes.
- clean up tests, using .loc instead of query to avoid issues with numexpr
- using `.loc` for tests also improves the speed of the tests.

### take with a pinch of salt

```py
df1 = pd.DataFrame({'id': [1,1,1,2,2,3], 'value_1': [2,5,7,1,3,4]})
df2 = pd.DataFrame({'id': [1,1,1,1,2,2,2,3], 'value_2A': [0,3,7,12,0,2,3,1], 'value_2B': [1,5,9,15,1,4,6,3]})

df1
   id  value_1
0   1        2
1   1        5
2   1        7
3   2        1
4   2        3
5   3        4

df2
   id  value_2A  value_2B
0   1         0         1
1   1         3         5
2   1         7         9
3   1        12        15
4   2         0         1
5   2         2         4
6   2         3         6
7   3         1         3

def outcome(sort):
        return df1.conditional_join(
                df2,
                ('value_1', 'value_2A', '>'),
                ('value_1', 'value_2B', '<'),
                ('id', 'id', '=='),
               how='inner',
               sort_by_appearance = sort
           )

In [28]:  %timeit df1.merge(df2, on='id').loc[lambda df: df.value_1.ge(df.value_2A) & df.value_1.lt(df.value_2B)]
2.85 s ± 14.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [31]: %memit df1.merge(df2, on='id').loc[lambda df: df.value_1.ge(df.value_2A) & df.value_1.lt(df.value_2B)]
peak memory: 4066.43 MiB, increment: 3638.79 MiB

# dev
In [26]: %timeit outcome(sort=True)
1.45 s ± 24.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [27]: %timeit outcome(sort=False)
1.31 s ± 17.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [32]: %memit outcome(sort=True)
peak memory: 1703.56 MiB, increment: 1357.85 MiB

In [33]: %memit outcome(sort=False)
peak memory: 1668.59 MiB, increment: 1373.60 MiB


# this PR
In [20]: %timeit outcome(sort=True)
498 ms ± 17.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [21]: %timeit outcome(sort=False)
381 ms ± 10.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [22]: %memit outcome(sort = True)
peak memory: 689.82 MiB, increment: 137.07 MiB

In [23]: %memit outcome(sort = False)
peak memory: 689.77 MiB, increment: 137.09 MiB
```
**This PR improves conditional_join.**

# PR Checklist

Please ensure that you have done the following:

1. [ ] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [ ] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [ ] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1058--pyjanitor.netlify.app

**Comment by ericmjl** [comment]
Cool stuff, @samukweku! Can I ask what the key improvements are? Could you also edit the PR message with a summary of those key improvements? I found that with the improvements spelled out, it helps focus the review a bit more, otherwise it's difficult to follow along.

File: janitor/functions/conditional_join.py [hunk 1]
```diff
@@ -417,13 +414,12 @@ def _less_than_indices(
     where `left_c` is less than
     (but not equal to) `right_c` are returned.
 
-    A tuple of integer indexes for left_c and right_c is returned.
+    if multiple_conditions is False, a tuple of integer indexes
```
Comment (ericmjl) [inline janitor/functions/conditional_join.py]: I think we don't have `multiple_conditions` as an argument here, is that right, @samukweku? If so, does this part of the docstring need updating?

**Review by ericmjl — APPROVED** [review]
Things look mostly good to me. Only one point about docstrings to address, but I will pre-approve the PR first.

File: janitor/functions/conditional_join.py (hunk 1 above)
Comment (samukweku) [reply janitor/functions/conditional_join.py:417]: I updated it.

**Review by thatlittleboy — APPROVED** [review]

**Pull request merged**